// @bundle-stats/ui/lib/components/bundle-assets/bundle-assets.container.js
"use strict";

function _typeof(obj) { "@babel/helpers - typeof"; if (typeof Symbol === "function" && typeof Symbol.iterator === "symbol") { _typeof = function _typeof(obj) { return typeof obj; }; } else { _typeof = function _typeof(obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; }; } return _typeof(obj); }

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.enhance = void 0;

var _get2 = _interopRequireDefault(require("lodash/get"));

var _react = require("react");

var _recompose = require("recompose");

var webpack = _interopRequireWildcard(require("@bundle-stats/utils/lib-esm/webpack"));

var _utils = require("@bundle-stats/utils");

var _withCustomSort = require("../../hocs/with-custom-sort");

var _withFilteredItems = require("../../hocs/with-filtered-items");

var _withSearch = require("../../hocs/with-search");

var _bundleAssets = require("./bundle-assets.constants");

function _getRequireWildcardCache() { if (typeof WeakMap !== "function") return null; var cache = new WeakMap(); _getRequireWildcardCache = function _getRequireWildcardCache() { return cache; }; return cache; }

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } if (obj === null || _typeof(obj) !== "object" && typeof obj !== "function") { return { "default": obj }; } var cache = _getRequireWildcardCache(); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj["default"] = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(Object(source), true).forEach(function (key) { _defineProperty(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

var addRowFlags = function addRowFlags(_ref) {
  var items = _ref.items;
  var updatedItems = items.map(function (item) {
    var runs = item.runs;
    var isEntry = runs.map(function (run) {
      return run && run.isEntry;
    }).includes(true);
    var isInitial = runs.map(function (run) {
      return run && run.isInitial;
    }).includes(true);
    var isChunk = runs.map(function (run) {
      return run && run.isChunk;
    }).includes(true);
    var isAsset = !(isEntry || isInitial || isChunk);
    return _objectSpread(_objectSpread({}, item), {}, {
      isEntry: isEntry,
      isInitial: isInitial,
      isChunk: isChunk,
      isAsset: isAsset
    });
  });
  return {
    items: updatedItems
  };
};

var getIsNotPredictive = function getIsNotPredictive(key, runs) {
  return runs.reduce(function (agg, current, index) {
    if (agg) {
      return agg;
    }

    if (index + 1 === runs.length) {
      return agg;
    }

    if (current && runs[index + 1] && current.delta !== 0 && key !== current.name && current.name === runs[index + 1].name) {
      return true;
    }

    return agg;
  }, false);
};

var addRowIsNotPredictive = function addRowIsNotPredictive(_ref2) {
  var items = _ref2.items;
  return {
    items: items.map(function (item) {
      return _objectSpread(_objectSpread({}, item), {}, {
        isNotPredictive: getIsNotPredictive(item.key, item.runs)
      });
    })
  };
};

var getRowFilter = function getRowFilter(filters) {
  return function (item) {
    if (filters[_utils.ASSET_FILTERS.CHANGED] && !item.changed) {
      return false;
    }

    if (!(filters["".concat(_utils.ASSET_ENTRY_TYPE, ".").concat(_utils.ASSET_FILTERS.ENTRY)] && item.isEntry || filters["".concat(_utils.ASSET_ENTRY_TYPE, ".").concat(_utils.ASSET_FILTERS.INITIAL)] && item.isInitial || filters["".concat(_utils.ASSET_ENTRY_TYPE, ".").concat(_utils.ASSET_FILTERS.CHUNK)] && item.isChunk || filters["".concat(_utils.ASSET_ENTRY_TYPE, ".").concat(_utils.ASSET_FILTERS.ASSET)] && item.isAsset)) {
      return false;
    }

    if (!filters["".concat(_utils.ASSET_FILE_TYPE, ".").concat((0, _utils.getFileType)(item.key))]) {
      return false;
    }

    return true;
  };
};

var getCustomSort = function getCustomSort(sortId) {
  return function (item) {
    if (sortId === _bundleAssets.SORT_BY_NAME) {
      return item.key;
    }

    if (sortId === _bundleAssets.SORT_BY_DELTA) {
      return (0, _get2["default"])(item, 'runs[0].deltaPercentage', 0);
    }

    if (sortId === _bundleAssets.SORT_BY_SIZE) {
      return (0, _get2["default"])(item, 'runs[0].value', 0);
    }

    return [!item.isNotPredictive, !item.changed, !item.isInitial, !item.isEntry, !item.isChunk, item.key];
  };
};

var enhance = (0, _recompose.compose)((0, _recompose.withProps)(function (_ref3) {
  var jobs = _ref3.jobs;
  var items = (0, _react.useMemo)(function () {
    return webpack.compareBySection.assets(jobs);
  }, [jobs]);

  var defaultFilters = _objectSpread(_objectSpread(_defineProperty({}, _utils.ASSET_FILTERS.CHANGED, (jobs === null || jobs === void 0 ? void 0 : jobs.length) > 1), (0, _utils.getAssetEntryTypeFilters)(true)), (0, _utils.getAssetFileTypeFilters)(true));

  var allEntriesFilters = _objectSpread(_objectSpread(_defineProperty({}, _utils.ASSET_FILTERS.CHANGED, false), (0, _utils.getAssetEntryTypeFilters)(true)), (0, _utils.getAssetFileTypeFilters)(true));

  return {
    items: items,
    totalRowCount: items.length,
    defaultFilters: defaultFilters,
    allEntriesFilters: allEntriesFilters
  };
}), // @TODO run both transformations in one pass
(0, _recompose.withProps)(addRowFlags), (0, _recompose.withProps)(addRowIsNotPredictive), (0, _withSearch.withSearch)(), (0, _withFilteredItems.withFilteredItems)(getRowFilter), (0, _withCustomSort.withCustomSort)({
  sortItems: _bundleAssets.SORT_BY,
  getCustomSort: getCustomSort
}));
exports.enhance = enhance;
