// @bundle-stats/ui/lib/layout/footer/footer.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Footer = void 0;

var _react = _interopRequireDefault(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _classnames = _interopRequireDefault(require("classnames"));

var _container = require("../../ui/container");

var _config = _interopRequireDefault(require("../../config.json"));

var _footerModule = _interopRequireDefault(require("./footer.module.css"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

function _objectWithoutProperties(source, excluded) { if (source == null) return {}; var target = _objectWithoutPropertiesLoose(source, excluded); var key, i; if (Object.getOwnPropertySymbols) { var sourceSymbolKeys = Object.getOwnPropertySymbols(source); for (i = 0; i < sourceSymbolKeys.length; i++) { key = sourceSymbolKeys[i]; if (excluded.indexOf(key) >= 0) continue; if (!Object.prototype.propertyIsEnumerable.call(source, key)) continue; target[key] = source[key]; } } return target; }

function _objectWithoutPropertiesLoose(source, excluded) { if (source == null) return {}; var target = {}; var sourceKeys = Object.keys(source); var key, i; for (i = 0; i < sourceKeys.length; i++) { key = sourceKeys[i]; if (excluded.indexOf(key) >= 0) continue; target[key] = source[key]; } return target; }

/* eslint-disable jsx-a11y/anchor-has-content */
var StandardLink = function StandardLink(_ref) {
  var to = _ref.to,
      source = _ref.source,
      props = _objectWithoutProperties(_ref, ["to", "source"]);

  return /*#__PURE__*/_react["default"].createElement("a", _extends({
    href: "".concat(_config["default"].siteUrl).concat(to, "?utm_source=").concat(source)
  }, props));
};
/* eslint-enable jsx-a11y/anchor-has-content */


StandardLink.propTypes = {
  /** Link to value */
  to: _propTypes["default"].string.isRequired,

  /** UTM source */
  source: _propTypes["default"].string.isRequired
};

var Footer = function Footer(_ref2) {
  var className = _ref2.className,
      children = _ref2.children,
      Link = _ref2.Link,
      source = _ref2.source;
  return /*#__PURE__*/_react["default"].createElement(_container.Container, {
    className: (0, _classnames["default"])(_footerModule["default"].root, className),
    as: "footer"
  }, children && /*#__PURE__*/_react["default"].createElement("div", {
    className: _footerModule["default"].content
  }, children), /*#__PURE__*/_react["default"].createElement("div", {
    className: _footerModule["default"].info
  }, /*#__PURE__*/_react["default"].createElement("div", {
    className: _footerModule["default"].navigation
  }, /*#__PURE__*/_react["default"].createElement("div", {
    className: _footerModule["default"].navGroup
  }, /*#__PURE__*/_react["default"].createElement("h3", {
    className: _footerModule["default"].navGroupTitle
  }, /*#__PURE__*/_react["default"].createElement(Link, {
    to: "/documentation",
    className: _footerModule["default"].navGroupTitleLink,
    source: source
  }, "Documentation")), /*#__PURE__*/_react["default"].createElement("nav", {
    className: _footerModule["default"].nav
  }, /*#__PURE__*/_react["default"].createElement(Link, {
    to: "/documentation/setup",
    className: _footerModule["default"].navLink,
    source: source
  }, "Setup"), /*#__PURE__*/_react["default"].createElement(Link, {
    to: "/documentation/metrics-and-data",
    className: _footerModule["default"].navLink,
    source: source
  }, "Metrics and data"), /*#__PURE__*/_react["default"].createElement(Link, {
    to: "/documentation/front-end-ops-resources",
    className: _footerModule["default"].navLink,
    source: source
  }, "FrontEnd-Ops Resources"), /*#__PURE__*/_react["default"].createElement(Link, {
    to: "/releases",
    className: _footerModule["default"].navLink,
    source: source
  }, "Release notes"), /*#__PURE__*/_react["default"].createElement(Link, {
    to: "/documentation/faq",
    className: _footerModule["default"].navLink,
    source: source
  }, "FAQ"), /*#__PURE__*/_react["default"].createElement("a", {
    href: "https://github.com/relative-ci/roadmap/issues",
    className: _footerModule["default"].navLink
  }, "Roadmap"))), /*#__PURE__*/_react["default"].createElement("div", {
    className: _footerModule["default"].navGroup
  }, /*#__PURE__*/_react["default"].createElement("h3", {
    className: _footerModule["default"].navGroupTitle
  }, "Tools"), /*#__PURE__*/_react["default"].createElement("nav", {
    className: _footerModule["default"].nav
  }, /*#__PURE__*/_react["default"].createElement("a", {
    href: "https://github.com/relative-ci/bundle-stats",
    className: _footerModule["default"].navLink,
    target: "_blank",
    rel: "noopener noreferrer",
    title: "In-depth bundle analyzer for webpack(bundle size, assets, modules, packages)"
  }, "Bundle Stats"), /*#__PURE__*/_react["default"].createElement("a", {
    href: "https://compare.relative-ci.com/?utm_source=".concat(source),
    className: _footerModule["default"].navLink,
    title: "Side by side comparison for webpack/lighthouse/browsertime stats"
  }, "Compare"), /*#__PURE__*/_react["default"].createElement("a", {
    href: "https://video-compare.relative-ci.com/?utm_source=".concat(source),
    className: _footerModule["default"].navLink,
    title: "Side by side video playing for webpagetest / browsertime recordings."
  }, "Video Compare"))), /*#__PURE__*/_react["default"].createElement("div", {
    className: _footerModule["default"].navGroup
  }, /*#__PURE__*/_react["default"].createElement("h3", {
    className: _footerModule["default"].navGroupTitle
  }, "Social"), /*#__PURE__*/_react["default"].createElement("nav", {
    className: _footerModule["default"].nav
  }, /*#__PURE__*/_react["default"].createElement("a", {
    href: "https://github.com/relative-ci",
    className: _footerModule["default"].navLink
  }, "Github"), /*#__PURE__*/_react["default"].createElement("a", {
    href: "https://twitter.com/Relative_CI",
    className: _footerModule["default"].navLink
  }, "Twitter")))), /*#__PURE__*/_react["default"].createElement("div", {
    className: _footerModule["default"].branding
  }, /*#__PURE__*/_react["default"].createElement(Link, {
    className: _footerModule["default"].copyright,
    to: "/",
    title: "Go to homepage",
    source: source
  }, "\xA9", " ".concat(new Date().getFullYear(), " ").concat(_config["default"].appName)))));
};

exports.Footer = Footer;
Footer.defaultProps = {
  className: '',
  children: null,
  source: 'web',
  Link: StandardLink
};
Footer.propTypes = {
  /** Adopted child class name */
  className: _propTypes["default"].string,

  /** Content */
  children: _propTypes["default"].node,

  /** UTM source */
  source: _propTypes["default"].string,

  /** Link component */
  Link: _propTypes["default"].elementType
};
