// @bundle-stats/utils/lib/webpack/extract/modules-packages-duplicate.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.extractModulesPackagesDuplicate = void 0;
const get_1 = __importDefault(require("lodash/get"));
const map_1 = __importDefault(require("lodash/map"));
const orderBy_1 = __importDefault(require("lodash/orderBy"));
const sum_1 = __importDefault(require("lodash/sum"));
const config_1 = require("../../config");
const utils_1 = require("../utils");
const extractModulesPackagesDuplicate = (webpackStats, currentExtractedData) => {
    const source = get_1.default(currentExtractedData, 'metrics.packages', {});
    const packagesByName = Object.entries(source).reduce((agg, [packageName, packageData]) => {
        const name = utils_1.getPackagePublicName(packageName);
        const existingPackageData = agg[name];
        return {
            ...agg,
            [name]: {
                ...existingPackageData,
                children: [
                    ...((existingPackageData === null || existingPackageData === void 0 ? void 0 : existingPackageData.children) || []),
                    {
                        name: packageName,
                        value: packageData.value,
                    },
                ],
            },
        };
    }, {});
    const { count, duplicatePackages } = Object.entries(packagesByName).reduce((agg, [name, data]) => {
        if (data.children.length === 1) {
            return agg;
        }
        return {
            count: agg.count + 1,
            duplicatePackages: [
                ...(agg.duplicatePackages || []),
                {
                    name,
                    value: sum_1.default(map_1.default(data.children, 'value')),
                    ...data,
                },
            ],
        };
    }, { count: 0, duplicatePackages: [] });
    if (!count) {
        return {
            metrics: {
                duplicatePackagesCount: {
                    value: count,
                },
            },
        };
    }
    const duplicatePackagesByName = orderBy_1.default(duplicatePackages, 'value', 'desc').reduce((agg, { name, ...duplicatePackageData }) => ({
        ...agg,
        [name]: {
            ...duplicatePackageData,
            children: orderBy_1.default(duplicatePackageData.children, 'value', 'desc'),
        },
    }), {});
    const data = Object.entries(duplicatePackagesByName).reduce((agg, [packageName, duplicatePackageData]) => ({
        ...agg,
        [packageName]: duplicatePackageData.children.map(({ name }) => name),
    }), {});
    return {
        insights: {
            duplicatePackages: {
                type: config_1.INSIGHT_WARNING,
                data,
            },
        },
        metrics: {
            duplicatePackagesCount: {
                value: count,
            },
        },
    };
};
exports.extractModulesPackagesDuplicate = extractModulesPackagesDuplicate;
